borsh/src/schema.rs

```diff
@@ -158,6 +158,150 @@ pub trait BorshSchema {
 
     /// Get the name of the type without brackets.
     fn declaration() -> Declaration;
+
+    /// Returns the largest possible size of serialised object.
+    ///
+    /// Returns `None` if the maximal size overflows `usize`.  This may happen
+    /// for nasted dynamically-sized types.  For example, notice that maximum
+    /// serialised size of `Vec<u8>` is `4 + u32::MAX` (four bytes for length
+    /// and then `u32::MAX` elements) and then maximum size of `Vec<Vec<u8>>` is
+    /// `4 + u32::MAX * (4 + u32::MAX)` which overflows `usize`.
+    ///
+    /// ```
+    /// use borsh::BorshSchema;
+    ///
+    /// assert_eq!(Some(0), <()>::max_serialized_size());
+    /// assert_eq!(Some(8), <usize>::max_serialized_size());
+    /// assert_eq!(Some(4 + 4294967295), <String>::max_serialized_size());
+    /// ```
+    fn max_serialized_size() -> Option<usize> {
```

Comment:
@mina86 without diving deep on this at once, can you come up with cases, when it might make sense to override this method?
Similarly to #204, if only the default implementation is supposed to be used everywhere, it'd be better to make this a helper method in `borsh/src/schema_helpers.rs`

Reply:
https://docs.rs/bounded-vec/latest/bounded_vec/ is a most obvious example.  It’s schema would be an sequence but since the type enforces element limit the actual number of elements will rarely if ever reach u32::MAX.

Reply:
Moved.

Reply:
@mina86 why did you decide to move it?

Reply:
`max_serialized_size` doesn’t call itself recursively so even though it would be possible to override it for `BoundedVec`, it’s not possible to override it for `Option<BoundedVec>`.  So the calculation must be done purely from information in the definitions.